### PR TITLE
Make sure regexp StringMatches can be compiled by Go and Envoy

### PR DESCRIPTION
**Pull request by esnible**
Resolves https://github.com/istio/istio/issues/23436

With this change, the validation webhook of Istiod, and `istioctl validate`, check all StringMatches for regexps that cannot be compiled.

Such regexps are rejected by validation.

Similar code has been running for the last two months by the analyzer.  https://github.com/istio/istio/pull/21868 .  This makes those changes absolute by rejecting any VirtualService with invalid regexps.

### CODE REVIEW

**Comment by esnible** [comment]
I added cherrypick/release-1.6, because this seems low-impact, and we have gotten two issues about it.  Feel free to remove the cherrypick if this should wait until 1.7.

File: pkg/config/validation/validation.go [hunk 1]
```diff
@@ -2085,6 +2093,20 @@ func validateHTTPRoute(http *networking.HTTPRoute, delegate bool) (errs error) {
...
+	if re == "" {
+		return nil
+	}
+
+	_, err := regexp.Compile(re)
+	if err == nil {
+		return nil
+	}
+
+	return fmt.Errorf("%s regexp invalid: %w", where, err)
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:2107]: Nit: maybe reference a doc or something indicating we use re2 regex?

Comment (esnible) [reply pkg/config/validation/validation.go:2107]: I put a `hold` on to address this.  The error messages from `regexp.Compile()` are nice.  I ran this using `istioctl validate -f galley/pkg/config/analysis/analyzers/testdata/virtualservice_regexes.yaml` and the messages are:
```
	* VirtualService//bad-match: uri regexp invalid: error parsing regexp: missing closing ]: `[A-Z`
	* VirtualService//ecma-not-v2: uri regexp invalid: error parsing regexp: invalid or unsupported Perl syntax: `(?!`
```

Should it say something like 
```
VirtualService//ecma-not-v2: "uri": invalid or unsupported Perl syntax: `(?!`; Istio uses RE2 style regex-based match (https://github.com/google/re2/wiki/Syntax)
```
?

**Comment by howardjohn** [comment]
That makes sense to me. The reason it may cause confusion is because its
not invalid regex, its invalid re2 regex. So a user may be confused that
their "correct" regex is rejected because its a different flavor of regex

**Pull request merged**